docs/architecture/psa-thread-safety.md

```diff
@@ -0,0 +1,275 @@
+Thread safety of the PSA subsystem
+==================================
+
+## Requirements
+
+### Backward compatibility requirement
+
+Code that is currently working must keep working. There can be an exception for code that uses features that are advertised as experimental; for example, it would be annoying but ok to add extra requirements for drivers.
+
+(In this section, “currently” means Mbed TLS releases without proper concurrency management: 3.0.0, 3.1.0, and any other subsequent 3.x version.)
+
+In particular, if you either protect all PSA calls with a mutex, or only ever call PSA functions from a single thread, your application currently works and must keep working. If your application currently builds and works with `MBEDTLS_PSA_CRYPTO_C` and `MBEDTLS_THREADING_C` enabled, it must keep building and working.
+
+As a consequence, we must not add a new platform requirement beyond mutexes for the base case. It would be ok to add new platform requirements if they're only needed for PSA drivers, or if they're only performance improvements.
+
+Tempting platform requirements that we cannot add to the default `MBEDTLS_THREADING_C` include:
+
+* Releasing a mutex from a different thread than the one that acquired it. This isn't even guaranteed to work with pthreads.
+* New primitives such as semaphores or condition variables.
+
```

Comment:
Based on our verbal discussion we could add the requirement about portability.

```suggestion

### Portability

If we need to add new threading primitives to the threading layer we should aim for maximum portability. Because of the [Backward compatibility requirement](#Backward-compatibility-requirement) we can't add this to the existing threading abstraction layer, we will need to introduce a new one with a new flag (eg. `PSA_CRYPTO_THREADING_C`).

For this reason we should take the new primitives from the C11 standard and make our API compatible with it.
```

Reply:
Added at to the “Destruction of a key in use” section, since this isn't a requirement for the basic project.